pkg/tfbridge/provider.go

```diff
@@ -309,12 +307,148 @@ func (p *Provider) GetSchema(ctx context.Context,
 
 // CheckConfig validates the configuration for this Terraform provider.
 func (p *Provider) CheckConfig(ctx context.Context, req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "CheckConfig is not yet implemented")
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.CheckConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.news", label),
+		KeepUnknowns: true,
+		SkipNulls:    true,
+		RejectAssets: true,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "CheckConfig failed because of malformed resource inputs")
+	}
+
+	// Strip out Pulumi-only config variables.
+	tfVars := make(resource.PropertyMap)
+	for k, v := range news {
+		if _, has := p.info.ExtraConfig[string(k)]; !has {
+			tfVars[k] = v
+		}
+	}
+
+	// First make a Terraform config map out of the variables. We do this before checking for missing properties
+	// s.t. we can pull any defaults out of the TF schema.
+	config, err := MakeTerraformConfig(nil, tfVars, p.config, p.info.Config, nil, p.configValues, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal config state")
+	}
+
+	if p.info.PreConfigureCallback != nil {
+		if err = p.info.PreConfigureCallback(news, config); err != nil {
+			return nil, err
+		}
+	}
+
+	// So we can provide better error messages, do a quick scan of required configs for this
+	// schema and report any that haven't been supplied.
+	var missingKeys []*pulumirpc.CheckFailure
+	for key, meta := range p.config {
+		if meta.Required && !config.IsSet(key) {
+			name := TerraformToPulumiName(key, meta, nil, false)
+			fullyQualifiedName := tokens.NewModuleToken(p.pkg(), tokens.ModuleName(name))
+
+			// TF descriptions often have newlines in inopportune positions. This makes them present
+			// a little better in our console output.
+			descriptionWithoutNewlines := strings.Replace(meta.Description, "\n", " ", -1)
+			missingKeys = append(missingKeys, &pulumirpc.CheckFailure{
+				Property: fullyQualifiedName.String(),
+				Reason:   descriptionWithoutNewlines,
+			})
+		}
+	}
+
+	if len(missingKeys) > 0 {
+		return &pulumirpc.CheckResponse{Inputs: req.GetNews(), Failures: missingKeys}, nil
+	}
+
+	// Perform validation of the config state so we can offer nice errors.
+	warns, errs := p.tf.Validate(config)
+	for _, warn := range warns {
+		if err = p.host.Log(ctx, diag.Warning, "", fmt.Sprintf("provider config warning: %v", warn)); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Wrap(multierror.Append(nil, errs...), "could not validate provider configuration")
+	}
+
+	return &pulumirpc.CheckResponse{Inputs: req.GetNews()}, nil
 }
```

Comment:
@pgavlin do we have a protocol description of the provider RPC contract anywhere?  I *think* it is right to move this validation from Configure into CheckConfig - but not certain of the guarantees and requirements around timing of these two calls.

Reply:
I don't think that this code can move wholesale. If this check is never performed in `Configure`, older versions of the CLI that do not call `CheckConfig` will no longer get any validation of provider config. We should figure out some approach that allows us to perform validation in `Configure` if the inputs were not validated by `CheckConfig`. I think we can achieve this by having `Configure` only perform validation if the [`args` field of `ConfigureRequest`](https://github.com/pulumi/pulumi/blob/master/sdk/proto/provider.proto#L78) is empty.